### PR TITLE
feat!: upgrade slf4j to 2.0.3

### LINKED ISSUE #249: No log is printed with Slf4j 2.x.x
## Description

I apply the slf4j version 2.0.3 in android project.
I tried to print log with debug, error level but nothing is printed. I check via logcat.

If I rollback to version 1.7.36, it works.

## Steps to reproduce

implementation("org.slf4j:slf4j-api:2.0.3")
implementation("com.github.tony19:logback-android:2.0.0")

## Environment

Android version : 9
Device : Huawei Nova 3i

### PR DESCRIPTION
**Pull request by compscidr**
Not sure if this is quite the right way of going about it, but was noticing if I updated slf4j to 2.0.x I wasn't getting logs showing up in Android any longer.

It looks like the only breaking change was in a few places `Marker` turned to `List<Marker>` so I updated throughout the code to take this wherever it was needed.

Tried it out against an android project and logging is now working again, so seems to work.

I have a feeling this will probably make things *not* work any longer with older slf4j though, so you may also not want to merge it, or make a different version for slf4j 2.x or something.

fix #249

### CODE REVIEW

**Comment by compscidr** [comment]
Hm actually, I must have been mistaken on my test, it appears it actually did not work, haha whoops.

**Comment by compscidr** [comment]
Okay, realized it was not working previously because it needed an implementation of `org.slf4j.impl.LoggerServiceProvider` which I have now added, and also added a file to the `META-INF/services` which tells the service loader of slf4j which implementation to try to load as a logging provider. Seems to be working now by building the jar with your script and including it in an android project using slf4j `2.0.3`

**Comment by sorinirimies** [comment]
Hi, @compscidr, do you need some further testing support? I was wondering if this PR could afterwards be merged as our project relies on this library as well.
I am also asking myself whether in general, this library is still maintained or not.

**Comment by compscidr** [comment]
> Hi, @compscidr, do you need some further testing support? I was wondering if this PR could afterwards be merged as our project relies on this library as well. I am also asking myself whether in general, this library is still maintained or not.

So far I've been using the changes from the PR using the jitpack build since I pushed them and it seems to be working well. It would be great to hear how others are making out with it. I did some pretty janky quickfixes to just make it api compatible - I'm curious about the markers and whether they will work well.

I think merging directly into this repo now will likely break backwards compatibility issues with pre-2.x slf4j, so it might take some careful thought if both are to be maintained together. I'm not sure on the status of this repo / whether it is being maintained (it seems like there were some recent pushes in the last year or so), but if its not, I can make a fork that's compatible with 2.x and start to maintain it. I'd likely drop older slf4j support entirely and focus only on the new stuff.

**Comment by tony19** [comment]
Thanks for your contribution!

I humbly admit that I've neglected this project, as I'm focused on other projects (personal and professional). I've been wanting to rewrite this library in Kotlin, especially to remove the cruft inherited from logback. I had started that endeavor a while ago, but got sidetracked.

I'll find time in the next few weeks to address some of the issues and PRs, and your PR does seem worthwhile.

**Comment by Sin4th** [comment]
Hey @tony19,
If this branch works and compatibility with slf4j 2.0.3 is established, would mind merging this PR? We would really like to use this library again in our Android project. 

Cheers

**Comment by compscidr** [comment]
> Just tested the implementation with: `implementation 'com.github.compscidr:logback-android:slf4f-2.0.3-SNAPSHOT'`
> 
> It still prints out:
> 
> `2022-11-03 18:38:10.225 4584-4584/com.random.package W/System.err: SLF4J: No SLF4J providers were found. 2022-11-03 18:38:10.226 4584-4584/com.random.package W/System.err: SLF4J: Defaulting to no-operation (NOP) logger implementation 2022-11-03 18:38:10.226 4584-4584/com.random.package W/System.err: SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.`
> 
> No logs in either appender. Seems to me, that the provider is still not loading properly. Was that part of your fix, @compscidr?

Hm that's weird, I have it working in a pretty complicated project atm - when I get a chance I'll see if I can make it work with a minimal example or something to verify there isn't some other interaction happening.

**Comment by Sin4th** [comment]
Thanks that would be great. 

Are you also using the following?
* Target SDK 33
* target & source-Compatibility VERSION_11

Also, appreciate your fast reply.

Cheers

**Comment by compscidr** [comment]
> Thanks that would be great.
> 
> Are you also using the following?
> 
> * Target SDK 33
> * target & source-Compatibility VERSION_11
> 
> Also, thanks for your fast reply.
> 
> Cheers

Okay here's a pretty minimal sample project with it working: https://github.com/compscidr/hello-kotlin-android - it does use target sdk 33 and source-compat VERSION_11

This is what it output to androids logcat:
```2022-11-04 14:44:13.107 22021-22021 LOGSTRING               com.example.myapplication            D  TEST MESSAGE
2022-11-04 14:44:13.107 22021-22021 c.e.m.MainActivity      com.example.myapplication            D  [main                ] logger test
2022-11-04 14:44:13.107 22021-22021 c.e.m.TestComponent     com.example.myapplication            D  [main                ] testComponent: someFunction()
2022-11-04 14:44:13.107 22021-22021 c.e.l.MyClass           com.example.myapplication            D  [main                ] In the display function!
```

And then it also logged to papertrail which I also often use alongside.

File: logback-android/src/main/java/ch/qos/logback/classic/util/LogbackMDCAdapter.java [hunk 1]
```diff
@@ -195,4 +198,24 @@ public void setContextMap(Map<String, String> contextMap) {
     // the newMap replaces the old one for serialisation's sake
     copyOnThreadLocal.set(newMap);
   }
+
+  @Override
+  public void pushByKey(String key, String value) {
```
Comment (tony19) [inline logback-android/src/main/java/ch/qos/logback/classic/util/LogbackMDCAdapter.java:203]: Why are the changes in this file are necessary for this PR?

File: logback-android/build.gradle [hunk 2]
```diff
@@ -63,13 +63,14 @@ dependencies {
     testImplementation 'com.icegreen:greenmail:1.5.9'
     testImplementation 'dom4j:dom4j:1.6.1'
     testImplementation 'org.easytesting:fest-assert:1.4'
-    testImplementation "org.slf4j:integration:${slf4jVersion}"
+    testImplementation "org.slf4j:integration:2.0.0-alpha1" // don't seem to have any stable 2.x releases yet
```
Comment (tony19) [inline logback-android/build.gradle:66]: Was this change intentional? This PR changes `slf4jVersion` to `2.0.3` in `gradle.properties`, but then hard-codes it here.

File: logback-android/build.gradle [hunk 3]
```diff
@@ -47,13 +47,13 @@ android {
...
-    testImplementation('junit:junit:4.12') {
+    testImplementation('junit:junit:4.13.2') {
         exclude(module: 'hamcrest-core')
     }
 
     testImplementation 'org.hamcrest:hamcrest-junit:2.0.0.0'
-    testImplementation 'org.robolectric:robolectric:4.3.1'
-    testImplementation 'org.mockito:mockito-core:2.23.4'
+    testImplementation 'org.robolectric:robolectric:4.8.1'
+    testImplementation 'org.mockito:mockito-core:4.7.0'
```
Comment (tony19) [inline logback-android/build.gradle:56]: I plan on updating all deps separately. If they're needed for this PR, please explain why.

File: logback-android/src/main/java/ch/qos/logback/core/net/AbstractSocketAppender.java [hunk 4]
```diff
@@ -186,7 +186,7 @@ protected void append(E event) {
     }
   }
 
-  private void connectSocketAndDispatchEvents() {
+  protected void connectSocketAndDispatchEvents() {
```
Comment (tony19) [inline logback-android/src/main/java/ch/qos/logback/core/net/AbstractSocketAppender.java:189]: Why are these scope changes necessary for this PR?

**Review by tony19 — CHANGES_REQUESTED** [review]

File: logback-android/build.gradle (hunk 2 above)
Comment (compscidr) [reply logback-android/build.gradle:66]: Not sure why I did that, will revert.

File: logback-android/src/main/java/ch/qos/logback/core/net/AbstractSocketAppender.java (hunk 4 above)
Comment (compscidr) [reply logback-android/src/main/java/ch/qos/logback/core/net/AbstractSocketAppender.java:189]: It may have been an artifact of me using the alt-enter method to implement interfaces using intellij, will review any scope changes and try reverting them all.

File: logback-android/src/main/java/ch/qos/logback/classic/util/LogbackMDCAdapter.java (hunk 1 above)
Comment (compscidr) [reply logback-android/src/main/java/ch/qos/logback/classic/util/LogbackMDCAdapter.java:203]: It looks like its because of changes to the MDCAdapter interface in slf4fj 2.0: https://www.slf4j.org/api/org/slf4j/spi/MDCAdapter.html

File: logback-android/build.gradle (hunk 2 above)
Comment (compscidr) [reply logback-android/build.gradle:66]: ah actually its because they seem to not have a matching version available: https://mvnrepository.com/artifact/org.slf4j/integration - you can see the latest they offer is `2.0.0-alpha1`

File: logback-android/src/main/java/ch/qos/logback/core/net/AbstractSocketAppender.java (hunk 4 above)
Comment (compscidr) [reply logback-android/src/main/java/ch/qos/logback/core/net/AbstractSocketAppender.java]: Looked more closely, I think there were some warnings or errors with tests previously. In any case, I've reverted these changes to scope throughout.

**Review by tony19 — DISMISSED** [review]

**Comment by tony19** [comment]
Thanks for the updates! I'll merge this in 3.x, since it's technically a breaking change (as you pointed out).

**Review by tony19 — APPROVED** [review]

**Comment by tony19** [comment]
@compscidr For some reason, this change breaks the build. I'm not sure why the CI failure didn't mark the PR. I'm reverting your changes for now in #296.
